bgpd/bgp_conditional_adv.c

```diff
@@ -0,0 +1,335 @@
+				 * the conditional advertise scanner process.
+				 * otherwise when route-map is also configured
+				 * on same peer, routes in advertise-map may not
+				 * be advertised as expected.
+				 */
+				if ((update_type == ADVERTISE)
+				    && subgroup_announce_check(dest, pi, subgrp,
+							       dest_p, &attr,
+							       true))
+					bgp_adj_out_set_subgroup(dest, subgrp,
+								 &attr, pi);
+				else {
+					/* If default originate is enabled for
+					 * the peer, do not send explicit
+					 * withdraw. This will prevent deletion
+					 * of default route advertised through
+					 * default originate.
+					 */
+					if (CHECK_FLAG(
+						    peer->af_flags[afi][safi],
+						    PEER_FLAG_DEFAULT_ORIGINATE)
+					    && is_default_prefix(dest_p))
+						break;
+
+					bgp_adj_out_unset_subgroup(
+						dest, subgrp, 1,
+						bgp_addpath_id_for_peer(
+							peer, afi, safi,
+							&pi->tx_addpath));
+				}
+			}
+		}
+	}
+}
+
+/* Handler of conditional advertisement timer event.
+ * Each route in the condition-map is evaluated.
+ */
+static int bgp_conditional_adv_timer(struct thread *t)
+{
+	afi_t afi;
+	safi_t safi;
+	int pfx_rcd_safi;
+	struct bgp *bgp = NULL;
+	struct peer *peer = NULL;
+	struct peer_af *paf = NULL;
+	struct bgp_table *table = NULL;
+	struct bgp_filter *filter = NULL;
+	struct listnode *node, *nnode = NULL;
+	struct update_subgroup *subgrp = NULL;
+	route_map_result_t ret;
+
+	bgp = THREAD_ARG(t);
+	assert(bgp);
+
+	thread_add_timer(bm->master, bgp_conditional_adv_timer, bgp,
+			 CONDITIONAL_ROUTES_POLL_TIME, &bgp->t_condition_check);
+
+	/* loop through each peer and advertise or withdraw routes if
+	 * advertise-map is configured and prefix(es) in condition-map
+	 * does exist(exist-map)/not exist(non-exist-map) in BGP table
+	 * based on condition(exist-map or non-exist map)
+	 */
+	for (ALL_LIST_ELEMENTS(bgp->peer, node, nnode, peer)) {
+		if (!CHECK_FLAG(peer->flags, PEER_FLAG_CONFIG_NODE))
+			continue;
+
+		if (peer->status != Established)
+			continue;
+
+		FOREACH_AFI_SAFI (afi, safi) {
+			if (strmatch(get_afi_safi_str(afi, safi, true),
+				     "Unknown"))
+				continue;
+
+			if (!peer->afc_nego[afi][safi])
+				continue;
+
+			/* labeled-unicast routes are installed in the unicast
+			 * table so in order to display the correct PfxRcd value
+			 * we must look at SAFI_UNICAST
+			 */
+			pfx_rcd_safi = (safi == SAFI_LABELED_UNICAST)
+					       ? SAFI_UNICAST
+					       : safi;
+
+			table = bgp->rib[afi][pfx_rcd_safi];
+			if (!table)
+				continue;
+
+			filter = &peer->filter[afi][safi];
+
+			if (!filter->advmap.aname || !filter->advmap.cname
+			    || !filter->advmap.amap || !filter->advmap.cmap)
+				continue;
+
+			if (!peer->advmap_config_change[afi][safi]
+			    && !peer->advmap_table_change)
+				continue;
+
+			if (BGP_DEBUG(update, UPDATE_OUT))
+				zlog_debug(
+					"%s: %s - advmap_table_change %d advmap_config_change[%d][%d] %d ",
```

Comment:
[%d][%d] seems not very handy, I think converting to a string would be better.

Reply:
Replaced afi/safi with string and printing the debug message only if the variable is set.